File: rxbinding/src/main/java/com/jakewharton/rxbinding2/view/ViewFocusChangeObservable.java
package com.jakewharton.rxbinding2.view;

import android.view.View;
import android.view.View.OnFocusChangeListener;
import com.jakewharton.rxbinding2.InitialValueObservable;
import io.reactivex.Observer;
import io.reactivex.android.MainThreadDisposable;

final class ViewFocusChangeObservable extends InitialValueObservable<Boolean> {
  private final View view;

  ViewFocusChangeObservable(View view) {
    this.view = view;
  }

  @Override protected void subscribeListener(Observer<? super Boolean> observer) {
    Listener listener = new Listener(view, observer);
    observer.onSubscribe(listener);
    view.setOnFocusChangeListener(listener);
  }

  @Override protected Boolean getInitialValue() {
    return view.hasFocus();
  }

  static final class Listener extends MainThreadDisposable implements OnFocusChangeListener {
    private final View view;
    private final Observer<? super Boolean> observer;

    Listener(View view, Observer<? super Boolean> observer) {
      this.view = view;
      this.observer = observer;
    }

    @Override public void onFocusChange(View v, boolean hasFocus) {
      if (!isDisposed()) {
        observer.onNext(hasFocus);
      }
    }

    @Override protected void onDispose() {
      view.setOnFocusChangeListener(null);
    }
  }
}

File: rxbinding/src/main/java/com/jakewharton/rxbinding3/view/ViewFocusChangeObservable.kt
@file:JvmName("RxView")
@file:JvmMultifileClass

package com.jakewharton.rxbinding3.view

import android.view.View
import android.view.View.OnFocusChangeListener
import androidx.annotation.CheckResult
import com.jakewharton.rxbinding3.InitialValueObservable
import io.reactivex.rxjava3.core.Observer
import io.reactivex.rxjava3.android.MainThreadDisposable

/**
 * Create an observable of booleans representing the focus of `view`.
 *
 * *Warning:* The created observable keeps a strong reference to `view`. Unsubscribe
 * to free this reference.
 *
 * *Warning:* The created observable uses [View.setOnFocusChangeListener] to observe
 * focus change. Only one observable can be used for a view at a time.
 *
 * *Note:* A value will be emitted immediately on subscribe.
 */
@CheckResult
fun View.focusChanges(): InitialValueObservable<Boolean> {
  return ViewFocusChangeObservable(this)
}

private class ViewFocusChangeObservable(
  private val view: View
) : InitialValueObservable<Boolean>() {

  override val initialValue: Boolean
    get() = view.hasFocus()

  override fun subscribeListener(observer: Observer<in Boolean>) {
    val listener = Listener(view, observer)
    observer.onSubscribe(listener)
    view.onFocusChangeListener = listener
  }

  private class Listener(
    private val view: View,
    private val observer: Observer<in Boolean>
  ) : MainThreadDisposable(), OnFocusChangeListener {

    override fun onFocusChange(v: View, hasFocus: Boolean) {
      if (!isDisposed) {
        observer.onNext(hasFocus)
      }
    }

    override fun onDispose() {
      view.onFocusChangeListener = null
    }
  }
}
